[flang][debug] Support assumed shape arrays.

abidh opened the pull request:
This PR generates dwarf to extract the information about the arrays from descriptor. The DWARF needs the offset of the fields like `lower_bound` and `extent`. To calculate them, we are adding up the sizes of the fields above. It seems to work ok for the 64-bit target. Will have to see if some changes are required for the 32-bits targets.

As we use data layout now, some tests needed to be adjusted to have a dummy data layout to avoid failure.

With this change in place, GDB is able show the assumed shape arrays correctly.

  subroutine ff(n, m, arr)
    integer n, m
    integer :: arr(:, :)
    print *, arr
    do i = 1, n
      do j = 1, m
        arr(j, i) = (i * 5) + j + 10
      end do
    end do
    print *, arr
  end subroutine ff

Breakpoint 1, ff (n=4, m=3, arr=...) at test1.f90:13
13          print *, arr
(gdb) p arr
$1 = ((6, 7, 8, 9) (11, 12, 13, 14) (16, 17, 18, 19))
(gdb) ptype arr
type = integer (4,3)
(gdb) c
Continuing.
 6 7 8 9 11 12 13 14 16 17 18 19

abidh committed:
[flang][debug] Support assume shape arrays.

This PR generates dwarf to extract the information about the arrays
from descriptor. To calculate the offset of the fields in the
descriptor, we are adding up the sizes of the fields above. It seems to
work ok for 64-bit target. Will have to see if some changes are
required for 32-bits targets.

As we use data layout now, some tests needed to be adjusted to have
a dummy data layout to avoid failure.

With this change in place, GDB is able show the assumed shape arrays
correctly.

  subroutine ff(n, m, arr)
    integer n, m
    integer :: arr(:, :)
    print *, arr
    do i = 1, n
      do j = 1, m
        arr(j, i) = (i * 5) + j + 10
      end do
    end do
    print *, arr
  end subroutine ff

Breakpoint 1, ff (n=4, m=3, arr=...) at test1.f90:13
13          print *, arr
(gdb) p arr
$1 = ((6, 7, 8, 9) (11, 12, 13, 14) (16, 17, 18, 19))
(gdb) ptype arr
type = integer (4,3)
(gdb) c
Continuing.
 6 7 8 9 11 12 13 14 16 17 18 19

tblah commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -22,6 +26,60 @@ namespace fir {
+  std::optional<mlir::DataLayout> dl =
+      fir::support::getOrSetDataLayout(module, /*allowDefaultLayout=*/true);
+  if (!dl)
+    mlir::emitError(module.getLoc(), "Missing data layout attribute in module");
```
nit: it would be better to return early after emitting the error so that we don't crash when dereferencing `dl`.

abidh replied:
Done.

tblah commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -22,6 +26,60 @@ namespace fir {
+  // My testing for a 32-bit shows that base_addr* is correctly given as a
+  // 32-bit entity. The index types are 64-bit so I am a bit uncertain how
+  // the alignment will effect the calculation of offsets in that case.
+
```
Maybe this needs an assertion to catch cases where there is padding added to the structure. See `ConvertFIRToLLVMPattern::computeBoxSize` (`FIROpPatterns.cpp`) for an example.

abidh replied:
As @jeanPerier commented, the risk has been pushed to getDescFieldTypeModel<> and not in my code. So I have removed the comments.

abidh committed:
Fix build error.

Add static to getModel() and constexpr to template specializations.

abidh committed:
Handle review comments.

1. Return if data layout is null.
2. Use getComponentOffset to calculate the offset of any field in the descriptor.
This was suggested by @jeanPerier.

abidh commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -22,6 +26,60 @@ namespace fir {
+
+  // dims
+  dimsSize =
+      dl->getTypeSizeInBits(getDescFieldTypeModel<kDimsPosInBox>()(context));
```
Thanks for your suggestions. Using the approach results in much cleaner code. I have updated the PR.

jeanPerier reviewed (approved):
LGTM, thanks

tblah reviewed (approved):
Thanks for this! Just a couple of nits

abidh committed:
Handle review comments.

abidh commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -23,63 +23,45 @@
+  // descriptors like lower_bound and extent for each dimension.
+  mlir::Type llvmDimsType = getDescFieldTypeModel<kDimsPosInBox>()(context);
+  dimsOffset = getComponentOffset<kDimsPosInBox>(*dl, context, llvmDimsType);
+  dimsSize = dl->getTypeSize(getDescFieldTypeModel<kDimsPosInBox>()(context));
```
Thanks for catching this. Fixed now.

abidh commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -23,63 +23,45 @@
+std::uint64_t getComponentOffset(const mlir::DataLayout &dl,
+                                 mlir::MLIRContext *context,
+                                 mlir::Type llvmFieldType) {
+  assert(DescriptorField > 0 && DescriptorField < 10);
```
Done.
